sdk/core/core-rest-pipeline/test/bearerTokenAuthenticationPolicy.spec.ts

```diff
@@ -399,3 +699,163 @@ class MockRefreshAzureCredential implements TokenCredential {
+    expectedResponseCode: 200,
+    expectedClaims: `Bearer authorization_uri="https://login.windows.net/", error="insufficient_claims"`,
+    encodedClaims: null,
+  },
+  {
+    testName: "no comma seperating the params",
+    challenge: `Bearer authorization_uri="https://login.windows.net/" error_description="ran into some error"`,
+    expectedResponseCode: 200,
+    expectedClaims: `Bearer authorization_uri="https://login.windows.net/" error_description="ran into some error"`,
+    encodedClaims: null,
+  },
+  {
+    testName: "Challenge with unexpected error",
+    challenge: `Bearer authorization_uri="https://login.windows.net/", error="invalid_token", claims="ey=="`,
+    expectedResponseCode: 200,
+    expectedClaims: `Bearer authorization_uri="https://login.windows.net/", error="invalid_token", claims="ey=="`,
+    encodedClaims: null,
+  },
+];
+
+type ChallengeType = "CAE" | "NonCAE" | "Success";
+interface Challenges {
+  testName: string;
+  challengeOrder: ChallengeType[];
+  shouldResolved: boolean;
+  numberOfGetTokenCalls: number;
+}
+// Number of getToken calls should be 1 + number of challenge handled to account for initial request
+const challengesOrderTestCases: Challenges[] = [
+  {
+    testName: "should handle CAE challenge after non-CAE challenge with custom handler",
+    challengeOrder: ["NonCAE", "CAE", "Success"],
+    shouldResolved: true,
+    numberOfGetTokenCalls: 3,
+  },
+  {
+    testName: "should handle at max 2 challenges with custom handler",
+    challengeOrder: ["NonCAE", "CAE", "NonCAE"],
+    shouldResolved: false,
+    numberOfGetTokenCalls: 3,
+  },
+  {
+    testName: "should not handle 2 CAE challenges",
+    challengeOrder: ["CAE", "CAE"],
+    shouldResolved: false,
+    numberOfGetTokenCalls: 2,
+  },
+  {
+    testName: "should not handle 2 non-CAE challenges with custom handler",
+    challengeOrder: ["NonCAE", "NonCAE"],
+    shouldResolved: false,
+    numberOfGetTokenCalls: 2,
+  },
+];
+
+// Brought over from azure-tools/test-utils-vitest/src/matrix.ts because we cannot depend on the library
+/**
+ * Takes a jagged 2D array and a function and runs the function with every
+ * possible combination of elements of each of the arrays
+ *
+ * For strong type-checking, it is important that the `matrix` have a strong
+ * type, such as a `const` literal.
+ *
+ * @param values - jagged 2D array specifying the arguments and their possible
+ *                 values
+ * @param handler - the function to run with the different argument combinations
+ *
+ * @example
+ * ```typescript
+ * matrix([
+ *     [true, false],
+ *     [1, 2, 3]
+ *   ] as const,
+ *   (useLabels: boolean, attempts: number) => {
+ *     // This body will run six times with the following parameters:
+ *     // - true, 1
+ *     // - true, 2
+ *     // - true, 3
+ *     // - false, 1
+ *     // - false, 2
+ *     // - false, 3
+ *   });
+ * ```
+ */
+function matrix<T extends ReadonlyArray<readonly unknown[]>>(
```

Comment:
There is a `matrix` in test-utils sdk. Can we not re-use that? @minhanh-phan

Reply:
Core has migrated to vitest so we need to use the vitest version of test-utils. It'll create a circular dependency so I'm copying it over instead.